src/validators/generator.rs

```diff
@@ -0,0 +1,138 @@
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+
+use crate::build_tools::SchemaDict;
+use crate::errors::ValResult;
+use crate::input::{GenericIterator, Input};
+use crate::questions::Question;
+use crate::recursion_guard::RecursionGuard;
+use crate::ValidationError;
+
+use super::{build_validator, BuildContext, BuildValidator, CombinedValidator, Extra, Validator};
+
+#[derive(Debug, Clone)]
+pub struct GeneratorValidator {
+    validator: Box<CombinedValidator>,
+    name: String,
+}
+
+impl BuildValidator for GeneratorValidator {
+    const EXPECTED_TYPE: &'static str = "generator";
+
+    fn build(
+        schema: &PyDict,
+        config: Option<&PyDict>,
+        build_context: &mut BuildContext,
+    ) -> PyResult<CombinedValidator> {
+        let schema: &PyAny = schema.get_as_req(intern!(schema.py(), "items_schema"))?;
+        let validator = Box::new(build_validator(schema, config, build_context)?);
+        let name = format!("{}[{}]", Self::EXPECTED_TYPE, validator.get_name());
+        Ok(Self { validator, name }.into())
+    }
+}
+
+impl Validator for GeneratorValidator {
+    fn validate<'s, 'data>(
+        &'s self,
+        py: Python<'data>,
+        input: &'data impl Input<'data>,
+        extra: &Extra,
+        slots: &'data [CombinedValidator],
+        recursion_guard: &'s mut RecursionGuard,
+    ) -> ValResult<'data, PyObject> {
+        let iterator = input.validate_iter()?;
+        let validator = InternalValidator {
+            validator: self.validator.clone(),
+            slots: slots.to_vec(),
+            data: extra.data.map(|d| d.into_py(py)),
+            field: extra.field.map(|f| f.to_string()),
+            strict: extra.strict,
+            context: extra.context.map(|d| d.into_py(py)),
+            recursion_guard: recursion_guard.clone(),
+        };
+        let v_iterator = ValidatorIterator { iterator, validator };
+        Ok(v_iterator.into_py(py))
+    }
+
+    fn get_name(&self) -> &str {
+        &self.name
+    }
+
+    fn ask(&self, question: &Question) -> bool {
+        self.validator.ask(question)
+    }
+
+    fn complete(&mut self, build_context: &BuildContext) -> PyResult<()> {
+        self.validator.complete(build_context)
+    }
+}
+
+#[pyclass]
+#[derive(Debug, Clone)]
+struct ValidatorIterator {
+    iterator: GenericIterator,
+    validator: InternalValidator,
+}
+
+#[pymethods]
+impl ValidatorIterator {
+    fn __iter__(slf: PyRef<'_, Self>) -> PyRef<'_, Self> {
+        slf
+    }
+
+    fn __next__(mut slf: PyRefMut<'_, Self>, py: Python) -> Option<PyObject> {
+        let validator_clone = &mut slf.validator.clone();
```

Comment:
This clone should not be necessary, but I can't work out how borrow splitting should work with `PyRefMut`???

Reply:
I think you can do it with a pattern, e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f1b28fb50009a6223a939065504aa37d

```suggestion
        let Self { validator, iterator, .. } = &mut *slf;
```

Reply:
that's worked, thanks so much.